### DISCUSSION TITLE
Significant increase in memory consumption after reloading configs via tasks/reload-entities

### DISCUSSION THREAD

**Discussion by schmity** [ldproxy/ldproxy#1063] [Setup] [answered]
We use tasks/reload-entities?ids=*&types=providers in the Admin-connector to reload new or changed ldproxy-configs. Usually configs change or few are added.

Monitoring ldproxy memory consumption, we observe it increases significantly on initial reload after startup.
In an instance with 336 APIs memory consumption increases from 1.4 GB to 2,3 GB.

Further calls to reload-entities still increases memory consumption, but much less (several MB per call).

Given that configs only changed or a few are added, memory consumption should not change significantly after reload via reload-entities.

From our logs: 

requests.log
NO_IP_LOGGING - - [29/Sep/2023:15:36:00 +0200] "POST /ldproxy_intranet/reload-entities?ids=*&types=providers HTTP/1.1" 200 - "-" "axios/0.26.1" X-Forwarded-For: "-" "-" "-" 1835262 22876633249664 37159690
NO_IP_LOGGING - - [29/Sep/2023:15:36:00 +0200] "POST /ldproxy_intranet/reload-entities?ids=stadtgruen&types=services HTTP/1.1" 200 - "-" "axios/0.26.1" X-Forwarded-For: "-" "-" "-" 3919716 22876633249664 254901

procmeminfo.log monitoring memory consumption
Fri 29 Sep 2023 03:35:01 PM CEST ldproxy_intranet PID: 1665356 Resident Memory used: (RSS) 1371756 kB
Fri 29 Sep 2023 03:40:01 PM CEST ldproxy_intranet PID: 1665356 Resident Memory used: (RSS) 2284980 kB 
Fri 29 Sep 2023 03:45:01 PM CEST ldproxy_intranet PID: 1665356 Resident Memory used: (RSS) 2285236 kB

**Comment by cportele** [comment, marked as answer]
I did some profiling with a configuration of ri.ldproxy.net (daraa and vineyards reduced to Features Part 1 and 2), but with 500 additional daraa and vineyards APIs, so in total 501 daraa and 501 vineyards APIs.

After startup, I triggered garbage collection, ran `./xtractl entity reload "*" -t "providers"`. After completion I triggered garbage collection.

Initial observations about the changes before/after the reload:

* The number of FeatureProviderSql instances did not change (1002).
* The number of ImmutableFeatureProviderSqlData instances increased by 50% (plus 501 to 1503). This seems to indicate that either the "old" daraa or vineyards data instances are still in memory? It is unclear why (and why 501 and not 0 or 1002).
* The number of FeatureSchema instances increased by 80% (plus ~309000 to ~690000).
* The number of SchemaSql instances did not change (~920000).
* Free memory in the docker container decreased by ~600MB.

I then reduced the configuration to a single daraa API. Before the reload there is a single ImmutableFeatureProviderSqlData instance that is referenced twice:

* from FeatureProviderSql;
* from EventSourcing.cache (referenced from EntityDataStoreImpl and others).

After the reload there are two ImmutableFeatureProviderSqlData instances, one referenced from FeatureProviderSql and a different one from EventSourcing.cache.

I can see in the debugger that the EventSourcing.cache reference is removed and a new ImmutableFeatureProviderSqlData instance is added to the cache.

In AbstractEntityFactory.updateInstance() the setData() method is not executed, because "no effective changes detected". This results in two instances of ImmutableFeatureProviderSqlData for the API, both with identical contents. 

A similar issue likely exists for other entity types.

@azahnen - We should convert this to a bug issue, if my analysis is correct.
